### PR TITLE
Read secrets for onboarding-token validation

### PR DESCRIPTION
**Pull request by Mrudraia**
This PR reads the secrets instead of reading the secrets from the volume mounts. 
whenever the new onboarding secrets are created, it takes more time to read the secrets from the volume mounts,
The user clicks the rotate onboarding  keys, the kubernetes still uses the old  public, private keys , the new keys are mounted later,  So this PR will read the secrets directly from the kubernetes secrets.

### CODE REVIEW

File: controllers/util/provider.go [hunk 1]
```diff
@@ -64,16 +79,52 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
...
+	privateSecret.Name = onboardingValidationPrivateKeySecretName
+	privateSecret.Namespace = operatorNamespace
+	err = cl.Get(ctx, types.NamespacedName{Name: storageClusterName, Namespace: operatorNamespace}, privateSecret)
+	if err != nil && !kerrors.IsNotFound(err) {
+		klog.Exitf("failed to get private secret: %v", err)
+	}
+
+	pubKeyBytes := privateSecret.Data["key"]
+
+	Block, _ := pem.Decode(pubKeyBytes)
```
Comment (rchikatw) [inline controllers/util/provider.go:103]: pem.Decode(privateSecret.Data["key"])

File: controllers/util/provider.go [hunk 2]
```diff
@@ -64,16 +79,52 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
...
 	if err != nil {
 		return nil, fmt.Errorf("failed to parse private key: %v", err)
 	}
 	return privateKey, nil
 }
+
+func newClient() (client.Client, error) {
+	klog.Info("Setting up k8s client")
+	scheme := runtime.NewScheme()
+	if err := v1.AddToScheme(scheme); err != nil {
```
Comment (rchikatw) [inline controllers/util/provider.go:114]: I think clientgoscheme already registers Kube native APIs. Let's try without the scheme.

**Comment by rchikatw** [comment]
Do you think it would be a good idea to remove the mounted secret from the ocs-operator? The mounted secret is used to create the default client. @leelavg

**Comment by leelavg** [comment]
> Do you think it would be a good idea to remove the mounted secret from the ocs-operator? The mounted secret is used to create the default client. @leelavg
- Sure, I see no point in having two implementations for the same requirement.

File: controllers/util/provider.go [hunk 3]
```diff
@@ -64,16 +79,52 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
...
 	if err != nil {
-		return nil, fmt.Errorf("failed to read private key: %v", err)
+		klog.Exitf("unable to get operator namespace: %v", err)
 	}
 
-	Block, _ := pem.Decode(pemString)
+	privateSecret := &corev1.Secret{}
+	privateSecret.Name = onboardingValidationPrivateKeySecretName
+	privateSecret.Namespace = operatorNamespace
+	err = cl.Get(ctx, types.NamespacedName{Name: storageClusterName, Namespace: operatorNamespace}, privateSecret)
```
[the remark this answers is not in the document]
Comment (Mrudraia) [reply controllers/util/provider.go]: storageClusterName is removed

File: controllers/util/provider.go (hunk 1 above)
Comment (Mrudraia) [reply controllers/util/provider.go]: resolved

File: controllers/util/provider.go [hunk 4]
```diff
@@ -64,16 +76,52 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
...
 	}
 
-	Block, _ := pem.Decode(pemString)
+	privateSecret := &corev1.Secret{}
+	privateSecret.Name = onboardingValidationPrivateKeySecretName
+	privateSecret.Namespace = operatorNamespace
+	err1 := cl.Get(ctx, types.NamespacedName{Name: onboardingValidationPrivateKeySecretName, Namespace: operatorNamespace}, privateSecret)
+	if err1 != nil {
+		return nil, fmt.Errorf("failed to get private secret: %v", err1)
+	}
```
Comment (nb-ohad) [inline controllers/util/provider.go:96]: 1. There should be no need to create a new client. In most cases this code is being used there is already a client that is used on the call sire of `GenerateOnboardingToken`. It is important to use that client instead of the new client for many reasons, some of which are:
- The client is a cached client and the object is already loaded
- The client is a mock/testing client used for testing and does not execute to a valid k8s API server

2. I am not sure you are taking the right approach by trying to load the secret here. This code path was and should be a compute-only code path, this means it will benefit us to not introduce any k8s-specific concerns into it.  My approach will be to load the secret outside of the `GenerateOnboardingToken` util and provide the private key payload an argument

**Review by nb-ohad — CHANGES_REQUESTED** [review]

File: controllers/util/provider.go (hunk 4 above)
Comment (Mrudraia) [reply controllers/util/provider.go]: 1) Used existing client on the call site on GenerateOnboardingToken, 

2) The secrets are loaded in a separate function and the private key is decoded and passed to the GenerateOnboardingToken

**Comment by rchikatw** [comment]
i see the rebase is not done properly. Can you rebase correctly? Basically, I see 2 commits and one of the commits is unrelated to your changes.

File: services/ux-backend/handlers/onboardingtokens/handler.go [hunk 5]
```diff
@@ -36,6 +33,8 @@ func handlePost(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours int
 	// When ContentLength is 0 that means request body is empty and
 	// storage quota is unlimited
 	var err error
+	var cl client.Client
```
Comment (leelavg) [inline services/ux-backend/handlers/onboardingtokens/handler.go:36]: where is the kubeclient initialized?

File: controllers/storagecluster/odfinfoconfig.go [hunk 6]
```diff
@@ -2,21 +2,21 @@ package storagecluster
...
 	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
+	"gopkg.in/yaml.v2"
 	corev1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/api/errors"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	ctrl "sigs.k8s.io/controller-runtime"
 	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
 	"sigs.k8s.io/controller-runtime/pkg/reconcile"
-	"sigs.k8s.io/yaml"
```
Comment (leelavg) [inline controllers/storagecluster/odfinfoconfig.go:19]: why is this pkg being removed and different pkg added?

File: controllers/util/provider.go [hunk 7]
```diff
@@ -10,17 +11,26 @@ import (
...
+)
+
+const (
+	// Name of existing private key which is used ocs-operator
+	onboardingValidationPrivateKeySecretName = "onboarding-private-key"
 )
 
 // GenerateOnboardingToken generates a token valid for a duration of "tokenLifetimeInHours".
 // The token content is predefined and signed by the private key which'll be read from supplied "privateKeyPath".
 // The storageQuotaInGiB is optional, and it is used to limit the storage of PVC in the application cluster.
```
Comment (leelavg) [inline controllers/util/provider.go:32]: pls change or remove irrelevant parts of the comments.

File: controllers/util/provider.go [hunk 8]
```diff
@@ -64,16 +74,28 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
 	return fmt.Sprintf("%s.%s", encodedPayload, encodedSignature), nil
 }
 
-func readAndDecodePrivateKey(privateKeyPath string) (*rsa.PrivateKey, error) {
-	pemString, err := os.ReadFile(privateKeyPath)
+func DecodePemKey(cl client.Client) (*rsa.PrivateKey, error) {
```
Comment (leelavg) [inline controllers/util/provider.go:77]: does this func needs to be exported?

File: controllers/util/provider.go (hunk 7 above)
Comment (Mrudraia) [reply controllers/util/provider.go]: removed the irrelevant comments

File: controllers/storagecluster/odfinfoconfig.go (hunk 6 above)
Comment (Mrudraia) [reply controllers/storagecluster/odfinfoconfig.go]: This change is not mine, It accured due to improper rebasing

File: services/ux-backend/handlers/onboardingtokens/handler.go [hunk 9]
```diff
@@ -36,6 +35,11 @@ func handlePost(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours int
 	// When ContentLength is 0 that means request body is empty and
 	// storage quota is unlimited
 	var err error
+	client, err := newClient()
+	if err != nil {
+		klog.Errorf("failed to create new client. %v", err)
+	}
```
Comment (leelavg) [inline services/ux-backend/handlers/onboardingtokens/handler.go:41]: you shouldn't be creating kubeclient for every request, create it in main and use that.

File: controllers/util/provider.go [hunk 10]
```diff
@@ -64,16 +73,28 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
...
+	operatorNamespace, err := GetOperatorNamespace()
+	if err != nil {
+		return nil, fmt.Errorf("unable to get operator namespace: %v", err)
+	}
+
+	privateSecret := &corev1.Secret{}
+	privateSecret.Name = onboardingValidationPrivateKeySecretName
+	privateSecret.Namespace = operatorNamespace
+
+	err = cl.Get(ctx, types.NamespacedName{Name: onboardingValidationPrivateKeySecretName, Namespace: operatorNamespace}, privateSecret)
```
Comment (leelavg) [inline controllers/util/provider.go:88]: you have already set ObjectMeta on privateSecret and could do
```suggestion
	err = cl.Get(ctx, client.ObjectKeyFromObject(privateSecret), privateSecret)
```

File: controllers/util/provider.go [hunk 11]
```diff
@@ -64,16 +73,28 @@ func GenerateOnboardingToken(tokenLifetimeInHours int, privateKeyPath string, st
 	return fmt.Sprintf("%s.%s", encodedPayload, encodedSignature), nil
 }
 
-func readAndDecodePrivateKey(privateKeyPath string) (*rsa.PrivateKey, error) {
-	pemString, err := os.ReadFile(privateKeyPath)
+func decodePemKey(cl client.Client) (*rsa.PrivateKey, error) {
+	klog.Info("Decoding the Pem key")
+	ctx := context.Background()
```
Comment (leelavg) [inline controllers/util/provider.go:78]: better to get the context from caller rather than creating one here everytime.

File: services/ux-backend/handlers/onboardingtokens/handler.go [hunk 12]
```diff
@@ -86,3 +90,19 @@ func handleUnsupportedMethod(w http.ResponseWriter, r *http.Request) {
 		klog.Errorf("failed write data to response writer: %v", err)
 	}
 }
+
+func newClient() (client.Client, error) {
```
Comment (leelavg) [inline services/ux-backend/handlers/onboardingtokens/handler.go:94]: this is a handler and possibly creation of client shouldn't be here.

File: controllers/util/provider.go (hunk 11 above)
Comment (nb-ohad) [reply controllers/util/provider.go:78]: 👍

File: services/ux-backend/handlers/onboardingtokens/handler.go (hunk 9 above)
Comment (nb-ohad) [reply services/ux-backend/handlers/onboardingtokens/handler.go:41]: @Mrudraia I believe main already creates a client and some other endpoints accept that client. Can you please follow that pattern

File: controllers/util/provider.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (Mrudraia) [reply controllers/util/provider.go]: Changed the GenerateOnboardingToken method, privateKey is passed to the function, no k8s query is done inside GenerateOnboardingToken.

File: services/ux-backend/handlers/onboardingtokens/handler.go (hunk 12 above)
Comment (Mrudraia) [reply services/ux-backend/handlers/onboardingtokens/handler.go]: Ack,  removed the code in handler.go and as suggested called from the main.

File: controllers/util/provider.go (hunk 10 above)
Comment (Mrudraia) [reply controllers/util/provider.go]: Ack

File: services/ux-backend/main.go [hunk 13]
```diff
@@ -82,3 +105,48 @@ func main() {
...
+	}
+	k8sClient, err := client.New(config, client.Options{Scheme: scheme})
+	if err != nil {
+		return nil, err
+	}
+
+	return k8sClient, nil
+}
+
+func readPrivateKey(cl client.Client) (*rsa.PrivateKey, error) {
```
Comment (rchikatw) [inline services/ux-backend/main.go:128]: Do not read the private key in the main.go file as it is only called once at the start. It will be a problem if the storage admin rotates the key and the private key changes.

**Comment by Mrudraia** [comment]
> @Mrudraia: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> ci/prow/ocs-operator-bundle-e2e-aws	[496fd5e](https://github.com/red-hat-storage/ocs-operator/commit/496fd5e9a67d489816cdb985c2d3d0ab5a860ae6)	[link](https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/red-hat-storage_ocs-operator/2715/pull-ci-red-hat-storage-ocs-operator-main-ocs-operator-bundle-e2e-aws/1831319608603185152)	unknown	`/test ocs-operator-bundle-e2e-aws`
> [Full PR test history](https://prow.ci.openshift.org/pr-history?org=red-hat-storage&repo=ocs-operator&pr=2715). [Your PR dashboard](https://prow.ci.openshift.org/pr?query=is:pr+state:open+author:Mrudraia).

/retest

File: services/ux-backend/main.go (hunk 13 above)
Comment (Mrudraia) [reply services/ux-backend/main.go]: Removed the reading of private key from main.go, and moved to utils.

File: services/ux-backend/handlers/onboardingtokens/handler.go (hunk 9 above)
Comment (Mrudraia) [reply services/ux-backend/handlers/onboardingtokens/handler.go]: Acknowledge

File: services/ux-backend/handlers/onboardingtokens/handler.go [hunk 14]
```diff
@@ -22,20 +19,26 @@ var unitToGib = map[string]uint{
...
-func handlePost(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours int) {
+func handlePost(w http.ResponseWriter, r *http.Request, tokenLifetimeInHours int, cl client.Client) {
 	var storageQuotaInGiB *uint
 	// When ContentLength is 0 that means request body is empty and
 	// storage quota is unlimited
 	var err error
+
+	privateKey, err := util.ReadPrivateKey(cl)
+	if err != nil {
+		fmt.Println("Failed to get private key", err)
```
Comment (rchikatw) [inline services/ux-backend/handlers/onboardingtokens/handler.go:39]: it should be of type http.Error

File: controllers/util/util.go [hunk 15]
```diff
@@ -5,6 +5,8 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const onboardingValidationPrivateKeySecretName = "onboarding-private-key"
```
Comment (rchikatw) [inline controllers/util/util.go:8]: its also defined in storageclient.go

File: controllers/util/provider.go [hunk 16]
```diff
@@ -5,22 +5,18 @@ import (
...
 	"fmt"
-	"os"
 	"time"
 
 	"github.com/google/uuid"
 	"github.com/red-hat-storage/ocs-operator/v4/services"
 )
 
 // GenerateOnboardingToken generates a token valid for a duration of "tokenLifetimeInHours".
-// The token content is predefined and signed by the private key which'll be read from supplied "privateKeyPath".
```
Comment (rchikatw) [inline controllers/util/provider.go:21]: you are updating the parameter type so I would suggest this comment should be update accordingly.

File: controllers/storagecluster/storagecluster_controller.go [hunk 17]
```diff
@@ -225,7 +225,7 @@ func (r *StorageClusterReconciler) SetupWithManager(mgr ctrl.Manager) error {
 		Owns(&appsv1.Deployment{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Owns(&corev1.Service{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
 		Owns(&corev1.ConfigMap{}, builder.MatchEveryOwner, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
-		Owns(&corev1.Secret{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Owns(&corev1.Secret{}, builder.MatchEveryOwner, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
```
Comment (rchikatw) [inline controllers/storagecluster/storagecluster_controller.go:228]: Why this change is needed.

Comment (Mrudraia) [reply controllers/storagecluster/storagecluster_controller.go:228]: This will trigger reconcilation  to all owners whenever their is change in secrets.

File: controllers/util/provider.go (hunk 16 above)
Comment (Mrudraia) [reply controllers/util/provider.go:21]: Ack

File: controllers/util/util.go (hunk 15 above)
Comment (Mrudraia) [reply controllers/util/util.go:8]: removed unused code

File: services/ux-backend/handlers/onboardingtokens/handler.go (hunk 14 above)
Comment (Mrudraia) [reply services/ux-backend/handlers/onboardingtokens/handler.go]: Ack

**Review by rewantsoni — CHANGES_REQUESTED** [review]

**Comment by leelavg** [comment]
Close this as a copy is raised already.

**Comment by Nikhil-Ladha** [comment]
Opened a new PR #2827
